arraycontext/impl/pytato/utils.py

```diff
@@ -0,0 +1,97 @@
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+from typing import Any, Dict, Set, Tuple, Mapping
+from pytato.array import SizeParam, Placeholder as BasePlaceholder
+from pytato.array import ShapeType, TagsType, Array, DataWrapper
+from pytato.transform import CopyMapper
+from pytools import UniqueNameGenerator
+
+
+class Placeholder(BasePlaceholder):
+    """
+    An equality-comparable :class:`pytato.Placeholder`.
+    """
+    def __init__(self, name: str,
+                 shape: ShapeType,
+                 dtype: np.dtype,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, shape, dtype, tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class Normalizer(CopyMapper):
+    """
+    Helper mapper for :func:`normalize_pt_expr`. Every
+    :class:`pytato.DataWrapper` is replaced with a deterministic copy of
+    :class:`Placeholder`.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self.bound_arguments: Dict[str, Any] = {}
+        self.vng = UniqueNameGenerator()
+        self.seen_inputs: Set[str] = set()
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        if expr.name is not None:
+            if expr.name in self.seen_inputs:
+                raise ValueError("Got multiple inputs with the name"
+                                 f"{expr.name} => Illegal.")
+            self.seen_inputs.add(expr.name)
+
+        name = self.vng("_actx_dw")
```

Comment:
* What guarantees the uniqueness of these names?
* If you're given a name (via `expr.name`), why not use it?

Reply:
> What guarantees the uniqueness of these names?

We use a `pytools.UniqueNameGenerator` to get these names, so must be unique.

> If you're given a name (via expr.name), why not use it?

More pairs of isomorphic DAGs normalized to the same DAG.

Reply:
> so must be unique.

I guess my question was "among what", and the answer I'm hearing is "only among themselves". Could you document that? (And that's OK because you're banning other named things.)

> More pairs of isomorphic DAGs normalized to the same DAG at a very minor cost.

Hmm... I see. Add a comment?

Reply:
Could you add a comment motivating the name normalization?

Reply:
Added.